akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -41,112 +54,166 @@ class AffinityPool(numThreads: Int, affinityGroupSize: Int, tf: ThreadFactory) e
 
   private def tryEnqueue(command: Runnable) = {
     // does a mod on the hash code to determine which queue to put task in
-    val queueIndex = Math.abs(command.hashCode()) & (numThreads - 1)
+    val queueIndex = Math.abs(command.hashCode()) & (parallelism - 1)
     workQueues(queueIndex).add(command)
   }
 
-  private def onWorkerExit(w: ThreadPoolWorker): Unit = {
-    locked(mainLock) {
+  /**
+   * Each worker should go through that method while terminating.
+   * In turn each worker is responsible for modifying the pool
+   * state accordingly. For example if this is the last worker
+   * and the queue is empty and we are in a ShuttingDown state
+   * the worker can transition the pool to ShutDown and attempt
+   * termination
+   *
+   * Furthermore, if this worker has experienced abrupt termination
+   * due to an exception being thrown in user code, the worker is
+   * responsible for adding one more worker to compensate for its
+   * own termination
+   *
+   */
+  private def onWorkerExit(w: ThreadPoolWorker, abruptTermination: Boolean): Unit =
+    locked(bookKeepingLock) {
       workers.remove(w)
+      if (workers.isEmpty && !abruptTermination && poolState >= ShuttingDown) {
+        poolState = ShutDown // transition to shutdown and try to transition to termination
+        attemptPoolTermination()
+      }
+      if (abruptTermination && poolState == Running)
+        addWorker(workers, w.q)
     }
+
+  override def execute(command: Runnable): Unit = {
+    if (command == null)
+      throw new NullPointerException
+    if (!(poolState == Running && tryEnqueue(command)))
```

Comment:
This is a bit tricky. The way ThreadPoolExecutor handles that is by rechecking the state after successful enqueue: 
```
        int c = ctl.get();
        if (isRunning(c) && workQueue.offer(command)) {
            int recheck = ctl.get();
            if (! isRunning(recheck) && remove(command))
                reject(command);
            else if (workerCountOf(recheck) == 0)
                addWorker(null, false);
        }
        else if (!addWorker(command, false))
            reject(command);
```

For starters we do not have a way to "remove" the task from the queue as this is a MPSC queue and that will totally not be tread safe. They use that facility to be able to report that the Runnable is rejected by throwing an exception. 

However, what happens if between isRunning and workQueue.offer a shutDown is called ? You will end up with a task in the queue that might be consumed by the time you recheck and try to remove it from the queue. 

So it seems to me (and I might be wrong) that they have relaxed some of the atomicity to avoid holding locks and thereby increase performance. 

Imagine holding a lock every time you enqueue... 

Any thoughts on that are more than welcome. I tried to think where to put synchronisation and where to avoid it.